### ISSUE TITLE
[Feature request] Python-like interface for modern C++

### ISSUE THREAD

**Issue by hcho3** [dmlc/xgboost#4895]
#4894 proposes to expose C++ headers to 3rd-party applications. As an extension, we should also consider building a Python-like frontend in modern C++. I'm thinking something like https://pytorch.org/cppdocs/frontend.html#end-to-end-example, where PyTorch offers a complete modern C++ API that's at feature parity with the Python package.

In particular, features like early stopping should be available from a C++ application.

**Comment by trivialfis** [comment]
Yeah, I love the interface from pytorch.  Too bad the torch script is based on Python.

**Comment by trivialfis** [comment]
It doesn't necessarily like Python though, we can improve upon it if needed.

**Comment by trams** [comment]
Looks good! I am all for moving common code from python|java|R to C++

**Comment by hcho3** [comment]
To new contributors: it would be great if we can have an experimental prototype for the C++ frontend. For the first pull request, it is okay to have only basic features (load data, train). Post comment here you'd like to work on this feature, so that we can discuss a possible design. I am available for help.

**Comment by hcho3** [comment]
I'd like to get to this issue, hopefully before end of this year.
